Local Environment Configuration and Lambda Logger Fixes

JStover95 opened the pull request:
# Local Environment Configuration and Lambda Logger Fixes

## Summary

This PR addresses local environment configuration improvements and enhances lambda logging functionality. The changes focus on security improvements, better error handling, and enhanced user feedback for lambda task operations.

## Changes

- Refactored logging formatter to properly handle datetime and date objects in ISO format, and added traceback information for error logging to improve debugging capabilities
- Cleaned up unused `FLASK_CONFIG` references from `EnvFileProvider` and `RootEnv` type definitions to allow overriding `Default` config with `Testing` or other configs locally.
- Updated init_lambda_task function to return task ID after creation and improved user feedback with success messages
- Implemented validation checks for `SECRET_KEY` and `JWT_SECRET_KEY` in `app.py` and updated `config.py` to retrieve these keys from environment variables for enhanced security

## How to Test

1. Verify that `SECRET_KEY` and `JWT_SECRET_KEY` both inherit the `FLASK_SECRET_KEY` env variable when `FLASK_CONFIG` is `Staging` or `Production`.
2. Verify that running the app with `FLASK_CONFIG` set to `Staging` or `Production` throws an error.
3. Test `flask init-lambda-task --status <Status>` to ensure task ID is returned and success messages are displayed.
5. Ensure all unit tests pass.

## Checklist

- [x] New code includes a license statement.
- [x] Code is properly linted (for Markdown, TypeScript, and Python).
- [x] All existing unit tests pass.

JStover95 committed:
Add IamSqlAlchemy class for IAM authentication with SQLAlchemy

- Introduce a new `IamSqlAlchemy` class that extends the SQLAlchemy functionality to support IAM authentication. The class includes methods for initializing the app with IAM options and generating authentication tokens using AWS RDS. This enhancement allows for secure database connections using IAM credentials, improving the overall security posture of the application.

JStover95 committed:
Refactor database initialization to support IAM authentication

- Update `register_extensions` in `app.py` to conditionally initialize the database with IAM support based on the environment (development/testing vs. production).
- Replace `SQLAlchemy` with `IamSqlAlchemy` in `extensions.py` to enhance security by integrating IAM authentication for database connections.

JStover95 committed:
Enhance IamSqlAlchemy for improved IAM SSL mode handling

- Introduced `SSLMode` type for `iam_sslmode` parameter in `init_app` method to enforce valid SSL mode values.
- Refactored the constructor to initialize the RDS client directly within the `IamSqlAlchemy` class, ensuring consistent client management.
- Removed redundant client initialization from `init_app` to streamline the database setup process.

JStover95 committed:
Add unit tests for IamSqlAlchemy class

- Introduced a new test file `test_iam_sqlalchemy.py` to validate the functionality of the `IamSqlAlchemy` class.
- Implemented tests for various scenarios including initialization with and without IAM, custom SSL modes, and exception handling during token generation.
- Ensured comprehensive coverage of the `init_app` method and its interaction with the database connection parameters.

JStover95 committed:
Enhance error handling in IamSqlAlchemy class

- Added logging for exceptions during IAM authentication token generation in the `IamSqlAlchemy` class.
- Introduced a logger to capture error messages, improving visibility into failures when creating authentication tokens.

JStover95 committed:
Add VPC CloudFormation template for Lambda and RDS infrastructure

- Introduced a new CloudFormation template `vpc-template.yaml` to define a VPC setup including public and private subnets, NAT instance, internet gateway, and security groups.
- Configured parameters for application name, VPC CIDR block, bastion AMI, and key pair for bastion host access.
- Implemented VPC endpoints for AWS services like Secrets Manager, RDS, and S3 to enhance security and connectivity.
- Added outputs for VPC and subnet IDs, as well as security group IDs for easy reference in other stacks.

JStover95 committed:
Add CloudFormation template for Aurora Serverless v2 PostgreSQL cluster

- Introduced a new CloudFormation template `rds-template.yaml` to define an Aurora Serverless v2 PostgreSQL cluster within a VPC.
- Configured parameters for application name, environment, database settings, and bootstrap configurations.
- Defined resources including DB cluster, subnet group, and bootstrap Lambda function with IAM roles for secure access.
- Added outputs for database cluster endpoint, port, and IAM user ARN for easy integration with other stacks.

JStover95 committed:
Add script for managing IP whitelisting in security groups

- Introduced `whitelist-bastion-ip.sh` to facilitate adding, removing, and listing IP addresses in a specified security group.
- Implemented validation functions for IP address and security group ID formats.
- Integrated AWS CLI commands for managing security group ingress rules with error handling.
- Provided usage examples and detailed comments for clarity and ease of use.

JStover95 committed:
Enhance VSCode settings with custom YAML tags

- Updated `.vscode/settings.sample.json` to include additional custom YAML tags for better syntax highlighting and validation in YAML files.
- Retained existing cSpell words while expanding the list to improve spell checking capabilities.

JStover95 committed:
Add database bootstrapper functionality with Docker support

- Introduced a new `DbUri` class for managing database URIs, including support for optional passwords.
- Added `Dockerfile` for building a Docker image with necessary dependencies for the database bootstrapper.
- Implemented `lambda_handler.py` to handle AWS Lambda events, including database upgrades and IAM user setup.
- Created `load_data.py` for loading data from JSON files into the database, with sequence management.
- Added scripts for running and testing the Docker image, ensuring a streamlined development process.
- Established backend extensions for Flask integration and database management.
- Included necessary files for a complete database bootstrapper setup, enhancing deployment capabilities.

JStover95 committed:
Update requirements.txt to include cfnresponse dependency

- Added `cfnresponse==1.1.5` to the requirements for enhanced CloudFormation response handling capabilities.

JStover95 committed:
Remove unused RDS stopping functionality and related database status checks

- Deleted `rds_stopper.py` which contained logic for stopping RDS instances based on inactivity.
- Removed `useDbStatus` hook from frontend components to streamline database connection handling.
- Updated login pages to eliminate dependency on database status checks during authentication.

JStover95 committed:
Add .pem file extension to .gitignore

- Updated the .gitignore file to include *.pem files, preventing sensitive key files from being tracked in the repository.

JStover95 committed:
Enhance environment handling by adding 'staging' type and updating conditional checks

- Expanded the `Environment` type in `environment.ts` to include a new 'staging' option.
- Updated various context files to include 'staging' in environment checks for fetching data, ensuring proper functionality in the staging environment.

JStover95 committed:
Refactor Docker setup by updating .dockerignore and Dockerfile

- Cleaned up .dockerignore by removing unnecessary entries, streamlining the build context.
- Updated Dockerfile to copy specific backend and shared directories, along with essential Python files for improved organization and clarity in the Docker image build process.

JStover95 committed:
Add Dockerfile for moto_proxy service

- Created a new Dockerfile to set up the moto_proxy service with Python 3.12.
- Installed necessary dependencies including boto3, flask, and moto for enhanced functionality.
- Configured the entry point for the service to run on host 0.0.0.0.

JStover95 committed:
Enhance DbUri class to URL-encode passwords

- Updated the `DbUri` class in `db_uri.py` to URL-encode passwords using `quote_plus`, ensuring special characters are handled correctly in database URIs.
- This change improves security and compatibility when constructing database connection strings.

JStover95 committed:
Implement local database support in lambda_handler.py

- Added a LOCAL_DB environment variable to toggle between local and IAM database connections in the `lambda_handler.py`.
- Updated the `create_app` function to conditionally initialize the app with or without IAM authentication based on the LOCAL_DB setting.
- Enhanced logging to indicate when the local database is being used and when no data is loaded.

JStover95 committed:
Refactor database bootstrapper script to support Aurora PostgreSQL

- Removed the creation of a dummy PostgreSQL database and replaced it with the setup of an Aurora PostgreSQL cluster and instance.
- Updated the script to retrieve database credentials from AWS Secrets Manager and use them for the local database setup.
- Enhanced the Docker run command to include the new database host and credentials.
- Added logic to wait for the database to be ready before proceeding with further operations.
- Cleaned up the script by ensuring proper stopping and removal of Docker containers after execution.

JStover95 committed:
Enhance test_image.sh to support dynamic request types

- Added command-line argument parsing to specify the RequestType (Create, Update, Delete).
- Implemented validation for the RequestType to ensure it is one of the allowed values.
- Updated the curl command to include the RequestType in the payload for function invocations, improving flexibility in testing scenarios.

JStover95 committed:
Refactor IAM SQLAlchemy tests to utilize environment-based database URI

- Updated `mock_app` fixture to use `FLASK_DB` environment variable for database URI configuration.
- Introduced `db_uri_params` fixture to extract and provide database connection parameters for tests.
- Modified test cases to assert against parameters derived from the environment, enhancing flexibility and security in test setups.

JStover95 committed:
Add sample Zappa settings configuration file

- Introduced a new `zappa_settings.sample.json` file to provide a template for Zappa deployment configurations.
- Included placeholders for essential settings such as AWS region, database URI, environment variables, and VPC configuration.
- This addition aids developers in setting up their Zappa projects with a clear reference for required parameters.

JStover95 committed:
Add installation guide for Ditti Research Dashboard</message>

<message>
- Introduced a new `INSTALL.md` file containing a comprehensive installation guide for the Ditti Research Dashboard.
- Included prerequisites, step-by-step installation instructions, and configuration details for VPC and RDS CloudFormation templates.
- Added usage information for the `whitelist-bastion-ip.sh` script and Zappa configuration settings.
- This guide aims to assist users in setting up and hosting their own version of the dashboard effectively.

JStover95 committed:
Update INSTALL.md to enhance clarity and structure

- Reformatted the installation steps with headings for better readability and organization.
- Added references to relevant CloudFormation templates and scripts for each step.
- Clarified the process for deploying the VPC, whitelisting IPs, setting up the database, and configuring Zappa.
- Improved instructions for deploying the Flask backend and Vite frontend, ensuring users have clear guidance throughout the installation process.

JStover95 committed:
Updated INSTALL.md

JStover95 committed:
Merge branch 'feature/JStover95/sqlalchemy-with-iam' into feature/JStover95/app-migration-to-vpc-and-aurora

JStover95 committed:
Add step to create dummy Zappa settings file in CI workflow

- Introduced a new step in the GitHub Actions workflow to create a dummy `zappa_settings.py` file.
- This addition facilitates the Docker image build process by ensuring the necessary Zappa configuration file is present.

JStover95 committed:
Refactor IP addition in whitelist-bastion-ip.sh to use ip-permissions format

- Updated the add_ip function to utilize the new --ip-permissions parameter for adding IPs to the security group.
- This change enhances the command's clarity and aligns with AWS best practices for security group modifications.

JStover95 committed:
Merge branch 'dev' into feature/JStover95/app-migration-to-vpc-and-aurora

JStover95 committed:
License

JStover95 committed:
Moved db bootstrapper source code to use src directory structure

JStover95 committed:
Update ruff.toml to refine linting rules for test files

- Modified the pattern for test files to include subdirectories and added new linting rules (S105, S106) to not check for hardcoded passwords in tests.

JStover95 committed:
Update VSCode settings to include additional pytest arguments for database bootstrapper tests

JStover95 committed:
Add unit tests for DbUri class

- Introduced new test cases for the DbUri class, covering various initialization scenarios including with and without passwords, handling special characters, and ensuring correct string representation.
- Added tests to validate port conversion and behavior with empty strings, enhancing test coverage for database URI generation.

JStover95 committed:
Add DataProcessor class and unit tests

- Introduced the DataProcessor class for handling data processing and transformation, including a method to clean row data by converting ISO format strings to datetime objects.
- Added unit tests for the DataProcessor class to validate the cleaning functionality, ensuring correct handling of valid and invalid datetime strings.
- Created an __init__.py file to organize utility modules within the db_bootstrapper package.

JStover95 committed:
Add FileReader class for JSON file reading

- Introduced the FileReader class to facilitate reading JSON files from the local filesystem.
- Implemented a method, read_json, that loads JSON data and returns it as a dictionary, enhancing data handling capabilities in the db_bootstrapper module.

JStover95 committed:
Add DbUri class for database URI management

- Introduced the DbUri class to facilitate the creation and management of database URIs, including support for username and password handling.
- Implemented methods for URI construction and string representation, ensuring secure handling of sensitive information.
- This addition enhances the db_bootstrapper module by providing a structured way to manage database connection details.

JStover95 committed:
Moved test_db_uri

JStover95 committed:
Add test fixtures and mock data for db_bootstrapper tests

- Introduced a new `conftest.py` file containing test fixtures for Flask application context and a mock file reader.
- Added `mock_data.json` to provide sample data for testing the mock database table.
- Created a `mock_file_reader.py` to facilitate loading mock data, enhancing the testing framework for the db_bootstrapper module.

JStover95 committed:
Add DatabaseSessionManager class and corresponding tests

- Introduced the DatabaseSessionManager class to manage database session operations, including methods for creating and closing sessions.
- Added unit tests for DatabaseSessionManager to validate session creation and closure, ensuring proper functionality within the db_bootstrapper module.

JStover95 committed:
Add unit tests for FileReader class

- Introduced a new test file for the FileReader class, validating its functionality for reading JSON files.
- Added tests for successful JSON reading, handling non-existent files, and managing invalid JSON content, enhancing test coverage for the db_bootstrapper module.

JStover95 committed:
Enhance test fixtures for db_bootstrapper with PostgreSQL support

- Introduced a MockPostgresContainer class to facilitate testing with a PostgreSQL database in a Docker container.
- Updated the test_client fixture to use the PostgreSQL container, ensuring a more realistic testing environment.
- Added a new MockEmptyTable class for additional mock database structure, improving test coverage and flexibility.

JStover95 committed:
Refactor mock file reader to use absolute path for loading mock data

- Updated the load_mock_data function in mock_file_reader.py to construct the file path using os.getcwd(), ensuring the correct location of mock_data.json regardless of the current working directory.
- This change enhances the reliability of the test setup for the db_bootstrapper module.

JStover95 committed:
Add SequenceManager class and unit tests for database sequence operations

- Introduced the SequenceManager class to manage database sequence operations, including methods for retrieving tables with ID columns, getting maximum IDs, checking sequence existence, resetting sequences, and fixing sequences.
- Added comprehensive unit tests for the SequenceManager class to validate its functionality, ensuring accurate handling of database sequences and enhancing test coverage for the db_bootstrapper module.

JStover95 committed:
Refactor test fixtures in conftest.py for improved PostgreSQL integration

- Replaced the mock_file_reader fixture with a new mock_postgres_container fixture to streamline PostgreSQL container usage in tests.
- Updated the test_client fixture to depend on the mock_postgres_container, enhancing the testing environment setup.
- Introduced with_mock_tables and with_mock_data fixtures to manage database schema and populate mock data, improving test reliability and coverage.
- Added alembic_version to mock_data.json for version tracking in tests.

JStover95 committed:
Refactor SequenceManager to use Enum for status messages

- Introduced SequenceManagerMessage Enum to standardize status messages related to sequence management.
- Updated SequenceManager methods to utilize the Enum for improved readability and maintainability of status messages.
- Adjusted unit tests to validate the new message format, ensuring consistency in expected outputs.

JStover95 committed:
Add DataLoader class and unit tests for data loading operations

- Introduced the DataLoader class to facilitate loading data from JSON files into the database, including error handling for missing tables and sequence management post-insertion.
- Added unit tests for the DataLoader class to validate its functionality, ensuring accurate data loading and enhancing test coverage for the db_bootstrapper module.
- Implemented status messages for various loading outcomes to improve user feedback during data operations.

JStover95 committed:
Add DatabaseManager class for IAM user management

- Introduced the DatabaseManager class to handle database operations related to IAM user setup, including user creation, permission grants, and connection testing.
- Added DatabaseManagerTextClause Enum for SQL query management, improving code readability and maintainability.
- Updated __init__.py to include DatabaseManager in the module's exports, ensuring accessibility for other components.

JStover95 committed:
Add unit tests for DatabaseManager class

- Introduced a new test file for the DatabaseManager class, implementing unit tests to validate IAM user setup, connection testing, and permission grants.
- Utilized mock objects to simulate database interactions, ensuring comprehensive coverage of the DatabaseManager's functionality.
- Enhanced test reliability by defining expected SQL calls and verifying their execution during tests.

JStover95 committed:
Add S3FileManager class for AWS S3 file operations

- Introduced the S3FileManager class to manage file downloads from AWS S3, including error handling for download failures.
- Updated __init__.py to export S3FileManager, ensuring accessibility for other components in the module.

JStover95 committed:
Add unit tests for S3FileManager class

- Introduced a new test file for the S3FileManager class, implementing unit tests to validate file download functionality, including both successful downloads and error handling.
- Utilized the moto library to mock AWS S3 interactions, ensuring isolated and reliable tests for the S3 file operations.

JStover95 committed:
Add SecretManager class for AWS Secrets Management

- Introduced the SecretManager class to interact with AWS Secrets Manager, allowing retrieval of secrets using their ARNs.
- Updated __init__.py to export SecretManager, ensuring it is accessible for other components in the module.

JStover95 committed:
Add unit tests for SecretManager class

- Introduced a new test file for the SecretManager class, implementing unit tests to validate secret retrieval functionality, including both successful retrieval and error handling.
- Utilized the moto library to mock AWS Secrets Manager interactions, ensuring isolated and reliable tests for the SecretManager operations.

JStover95 committed:
Add AppFactory class for Flask application creation

- Introduced the AppFactory class to facilitate the creation of Flask applications with database configuration.
- Implemented a static method to create a Flask app, integrating SQLAlchemy and migration support.
- Updated __init__.py to export AppFactory, ensuring accessibility for other components in the module.

JStover95 committed:
Add unit tests for AppFactory class

- Introduced a new test file for the AppFactory class, implementing unit tests to validate Flask app creation with proper database configuration.
- Verified that the SQLAlchemy database URI and tracking modifications are correctly set, ensuring the app is initialized as expected.

JStover95 committed:
Add DBBootstrapperAgent class for database bootstrapping operations

- Introduced the DBBootstrapperAgent class to handle CloudFormation requests for creating, updating, and deleting database configurations.
- Implemented methods for validating environment variables, retrieving database secrets, and creating Flask applications with appropriate database credentials.
- Added functionality for upgrading the database and setting up IAM authentication, along with data loading capabilities from S3.
- Included detailed logging for each operation to enhance traceability and error handling during the bootstrap process.

JStover95 committed:
Add equality and representation methods to DbUri class

- Implemented __eq__ method to allow comparison between DbUri instances based on their URI.
- Added __repr__ method for improved string representation of DbUri objects, including the URI and password visibility status.

JStover95 committed:
Refactor SQL query formatting in DatabaseManagerTextClause

- Updated SQL queries in DatabaseManagerTextClause to use Python's format method for variable substitution, enhancing readability and maintainability.
- Introduced a MIGRATION_DIR attribute in the DatabaseManager class to specify the migration directory during database upgrades.
- Adjusted the upgrade_database method to utilize the new MIGRATION_DIR attribute.
- Improved IAM user permission grants by formatting SQL queries directly with the iam_username variable.

JStover95 committed:
Enhance test setup for database bootstrapping

- Added DatabaseManager import to the test configuration for better integration.
- Defined IAM_USERNAME and MOCK_FILENAME constants for improved clarity in tests.
- Set the MIGRATION_DIR in DatabaseManager to specify the migration path for tests.
- Changed the test fixture scope from "module" to "session" for broader test context.
- Implemented a method to create a dummy role in the mock PostgreSQL container to facilitate role-based testing.

JStover95 committed:
Add unit tests for DBBootstrapperAgent class

- Introduced a comprehensive test suite for the DBBootstrapperAgent class, covering initialization, environment validation, and request handling for create, update, and delete operations.
- Implemented mock AWS services using the moto library to simulate interactions with Secrets Manager and S3, ensuring isolated testing.
- Validated the integration of database migrations and IAM user creation within the test scenarios, enhancing the robustness of the bootstrap process.
- Included tests for error handling and edge cases to ensure reliability in various operational contexts.

JStover95 committed:
Refactor lambda_handler for improved database bootstrapping

- Replaced the previous implementation of the lambda_handler function with a streamlined version that utilizes the DBBootstrapperAgent class for handling CloudFormation events.
- Removed unused imports and functions related to Flask and IAM user setup, simplifying the code structure.
- Added local database handling logic and improved error handling for better clarity and response management.
- Enhanced logging for success and failure scenarios during the bootstrap process.

JStover95 committed:
Add unit tests for lambda_handler function

- Created a new test file for the lambda_handler function, implementing unit tests to validate its behavior under different scenarios, including success, local database handling, and error conditions.
- Utilized mocking to simulate interactions with the DBBootstrapperAgent and cfnresponse, ensuring isolated testing of the lambda_handler logic.
- Enhanced test coverage to improve reliability and maintainability of the bootstrap process.

JStover95 committed:
Removed unused load_data.py

JStover95 committed:
Refactor message handling in DBBootstrapperAgent and related classes

- Moved message definitions from enums to a new enums module, enhancing organization and clarity.
- Updated message retrieval methods to use class methods for better encapsulation and readability.
- Adjusted tests to reflect changes in message handling, ensuring consistency across the codebase.
- Improved overall code maintainability by reducing direct string usage in favor of structured message methods.

JStover95 committed:
Refactor request handling in DBBootstrapperAgent

- Updated handle_create_request, handle_update_request, and handle_delete_request methods to remove unused event parameters, simplifying their signatures.
- Adjusted the main request handling logic to call the refactored methods without passing the event, enhancing code clarity and maintainability.

JStover95 committed:
Add DbConnectionExecutioner for database user management

- Introduced the DbConnectionExecutioner class to encapsulate database user management operations, including user creation and permission grants.
- Updated DatabaseManager to utilize DbConnectionExecutioner for improved clarity and maintainability in IAM user setup.
- Removed the DatabaseManagerTextClause class to streamline SQL query handling, directly integrating SQL commands within DbConnectionExecutioner.
- Enhanced error handling and connection management in DatabaseManager methods for better reliability during IAM user operations.

JStover95 committed:
Renamed enums.py to messages.py

JStover95 committed:
Renamed DbConnectionExecuter

JStover95 committed:
Refactor database connection utilities and enhance testing

- Updated import paths for DbConnectionExecuter to improve clarity and maintainability.
- Introduced mock connection and result classes in tests to facilitate unit testing of database operations.
- Added comprehensive tests for DbConnectionExecuter methods, ensuring correct handling of user management operations and SQL query validation.
- Streamlined DatabaseManager tests to utilize the updated DbConnectionExecuter, enhancing test coverage and reliability.

JStover95 committed:
Add shared package paths to test configuration

- Updated the conftest.py file to include parent directories in the Python path, allowing for the discovery of shared packages during testing.
- This change enhances the test environment setup, ensuring that necessary modules are accessible for the test suite.

JStover95 committed:
Refactor database connection executer methods for improved SQL command handling

- Updated methods in DbConnectionExecuter to use formatted SQL strings instead of parameterized queries for better readability and maintainability.
- Adjusted corresponding unit tests to reflect changes in SQL command formatting, ensuring accurate assertions on mock connection calls.
- Enhanced test coverage for user management operations, validating the correct execution of SQL commands in various scenarios.

JStover95 committed:
Update test configuration in conftest.py for improved fixture management

- Changed the mock_postgres_container fixture to be autouse, ensuring it is automatically used in tests without explicit declaration.
- Refactored the test_client fixture to remove dependency on mock_postgres_container, simplifying its definition while maintaining functionality.

JStover95 committed:
Refactor test_db_bootstrapper_agent to simplify request handling and improve fixture management

- Updated mock_migrations_dir fixture to yield the migrations directory for better test setup.
- Removed unused event parameters from handle_create_request, handle_update_request, and handle_delete_request methods in tests, enhancing clarity.
- Adjusted integration tests to reflect changes in request handling, ensuring accurate assertions on database state after migrations.

JStover95 committed:
Enhance test fixtures and mock implementations for AWS services

- Added mock fixtures for AWS Secrets Manager and S3 to streamline testing of database operations.
- Refactored test_db_bootstrapper_agent to utilize new mock fixtures, improving clarity and reducing redundancy.
- Updated test assertions to use constants for secret and data ARNs, enhancing maintainability and readability.
- Removed outdated mock implementations from tests, ensuring a cleaner and more efficient test suite.

JStover95 committed:
Updated Dockerfile COPY to match new patterns

JStover95 committed:
Refactor import statements in db_bootstrapper_agent, data_loader, and sequence_manager modules

- Updated import paths to use the new structure from src.utils.messages
- Added a print statement in lambda_handler to log received events, aiding in debugging and monitoring.

JStover95 committed:
Refactor testing utilities to ensure proper cleanup of mocks

- Updated the mock_boto3_client function to include a finally block for stopping the patcher, ensuring that mocks are properly cleaned up after use.
- Simplified pytest arguments in settings.sample.json by removing unnecessary flags, streamlining the testing configuration.

JStover95 committed:
Update test_backend.yml to specify test directories for pytest execution

- Modified the pytest command to run tests specifically in the 'tests' and 'functions/db_bootstrapper/tests_db_bootstrapper' directories, improving test organization and execution clarity.

JStover95 committed:
Removed inline noqa for S105 and S106 in favor of defining ignores in roff.toml

JStover95 committed:
Fixed package vulnerrabiliteis

JStover95 committed:
Merge branch 'fix/JStover95/npm-audit-fix' into feature/JStover95/app-migration-to-vpc-and-aurora

JStover95 committed:
Added AWS_DEFAULT_REGION to environ patch for failing test

JStover95 committed:
License

JStover95 committed:
Add IAM roles and policies for Aurora DB bootstrap Lambda

- Introduced a new CloudFormation template (iam-template.yaml) for defining IAM roles and policies required for the Aurora DB bootstrap Lambda function.
- Updated rds-template.yaml to import the IAM role from the new template, ensuring proper access to Secrets Manager and RDS.
- Removed inline IAM role definitions and replaced them with imported values for better modularity and maintainability.

JStover95 committed:
Refactor rds-template.yaml to remove StackIdentifier parameter and related references

- Eliminated the StackIdentifier parameter from the CloudFormation template.
- Updated DBClusterIdentifier, LogGroup, and resource tags to remove dependencies on StackIdentifier, simplifying the template structure.

JStover95 committed:
Refactor IAM and RDS templates for improved modularity

- Added metadata for parameter labels in iam-template.yaml to enhance clarity.
- Removed BootStrapDataARN parameter and its associated condition from iam-template.yaml.
- Moved the DBBootStrapS3AccessPolicyAttachment to rds-template.yaml, linking it to the imported IAM role for better resource management.
- Updated references to ensure proper integration between IAM and RDS templates.

JStover95 committed:
Fix importing lambda execution role from iam stack

JStover95 committed:
Renamed yml files

JStover95 committed:
Add port configuration to bootstrapper and retrieval scripts

- Introduced a new --port argument to allow dynamic port assignment in run_image.sh and test_image.sh for both db_bootstrapper and wearable_data_retrieval.
- Updated the curl commands to use the specified PORT variable instead of a hardcoded value, enhancing flexibility for local testing.

JStover95 committed:
Update whitelist-bastion-ip.sh and documentation for NAT security group

- Modified the script and documentation to specify 'nat-security-group-id' instead of 'security-group-id' for clarity.
- Updated usage examples and descriptions to reflect the changes in the security group terminology.

JStover95 committed:
Add BlockDeviceMappings to VPC template for EBS configuration

- Added BlockDeviceMappings section to specify EBS volume settings, including size, type, deletion on termination, and encryption for the root device in the VPC CloudFormation template.

JStover95 committed:
Add timeout configuration to Zappa settings sample

- Added a new "timeout_seconds" parameter set to 60 in the zappa_settings.sample.json file to account for Aurora spin-up time

JStover95 committed:
Add storage encryption to RDS configuration in CloudFormation template

- Enabled storage encryption for the RDS instance by setting StorageEncrypted to true in the rds-template.yml file, enhancing data security.

JStover95 committed:
Add secret key validation and environment variable configuration in app and config files

- Implemented a check in app.py to ensure SECRET_KEY and JWT_SECRET_KEY are set, raising an error if not.
- Updated config.py to retrieve SECRET_KEY and JWT_SECRET_KEY from environment variables for enhanced security and flexibility.

JStover95 committed:
Enhance lambda task initialization with status and return ID

- Updated the init_lambda_task function to return the task ID after creation, improving feedback for users.
- Modified the init-lambda-task command to display a success message with the initialized task ID.
- Adjusted the status option to remove the default value for clarity.

JStover95 committed:
Remove FLASK_CONFIG from environment variables in EnvFileProvider and RootEnv type definition

JStover95 committed:
Enhance JSON logging in lambda_logger.py

- Refactored the logging formatter to include a new method for converting datetime and date objects to ISO format, improving the handling of these types in log entries.
- Added functionality to print traceback information when logging errors, aiding in debugging.

JStover95 committed:
Update config.py to retrieve SECRET_KEY and JWT_SECRET_KEY from environment variables

- Added retrieval of SECRET_KEY and JWT_SECRET_KEY from environment variables to enhance security and flexibility in the Staging configuration.

JStover95 committed:
Merge branch 'dev' into feature/JStover95/local-env-config-and-lambda-logger-fixes

chris01b commented on backend/commands.py:
`LambdaTask.status` is a strict Enum (`Pending|InProgress|Success|Failed|CompletedWithErrors, nullable=False`). Passing `--status` absent/invalid will blow up at commit.

Maybe something like this?

```python
VALID_STATUSES = ["Pending", "InProgress", "Success", "Failed", "CompletedWithErrors"]

@click.command("init-lambda-task", help="Initialize a lambda task with the specified status.")
@click.option(
    "--status",
    required=True,
    type=click.Choice(VALID_STATUSES, case_sensitive=True),
    help="Status of the lambda task.",
)
@with_appcontext
def init_lambda_task_click(status: str):
    task_id = init_lambda_task(status)
    click.echo(f"Lambda task successfully initialized. ID: {task_id}")
```

chris01b commented on shared/lambda_logger.py:
`_convert_datetime` is annotated `-> str` but returns mixed. I think `Any` works best.

JStover95 committed:
Merge branch 'dev' into feature/JStover95/local-env-config-and-lambda-logger-fixes

chris01b commented:
TestDBBootstrapperAgentIntegration fails in this PR but passes in feature/JStover95/functions-migration-to-vpc-and-aurora-db-bootstrapper

JStover95 committed:
Add TaskStatusTypeEnum and refactor status column in LambdaTask model

- Introduced a new `TaskStatusTypeEnum` enum to standardize task status values, improving code clarity and maintainability.
- Refactored the `status` column in the `LambdaTask` model to utilize the new enum, enhancing type safety and consistency across the application.

JStover95 committed:
Refactor init-lambda-task command to use TaskStatusTypeEnum

- Updated the `init_lambda_task_click` function to accept a `TaskStatusTypeEnum` type for the status option, enhancing type safety and clarity.

JStover95 commented on backend/commands.py:
I created a `TaskStatusTypeEnum` in models.py that is now shared between the lambda task table and click command like you recommend. Refactoring the enum definition doesn't change the database schema. Running `flask db migrate` detects no changes in the schema and all unit tests pass.

JStover95 committed:
_convert_datetime annotated to return Any
